### PR TITLE
Port TH cum{sum,prod}_cuda to ATen

### PR DESCRIPTION
**Pull request by kshitij12345**
References: #24521 #24522 #24547 #24548 #24507

Depends on #36308 

Changes related to this PR are only in file : 
aten/src/ATen/Declarations.cwrap
aten/src/ATen/native/cuda/ReduceOpsKernel.cu
aten/src/ATen/native/native_functions.yaml
aten/src/THC/generic/THCTensorMathScan.cu
aten/src/THC/generic/THCTensorMathScan.h

Please Review @VitalyFedyunin 

Thanks.

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Hi! Can you please add perf comparison between master and this branch.

**Comment by kshitij12345** [comment]
Sure will do that and update.

**Comment by kshitij12345** [comment]
**Operator : cumsum**
Benchmark with same build settings on same system.
gcc : version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)
CUDA : 10.1
GPU : 1050ti

```python
import timeit

for n, t in [(500, 20000),
             (700, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.cumsum(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.cumsum(a, 0);torch.cumsum(a, 1); torch.cuda.synchronize()',
                              setup=f'import torch; a=torch.randn(({n}, {n}), dtype={dtype}, device="cuda")',
                              number=t))
```

Before:

```
torch.cumsum(a) a.numel() == 500 for 20000 times torch.half
2.693647157999976
torch.cumsum(a) a.numel() == 500 for 20000 times torch.float
3.7359478979999494
torch.cumsum(a) a.numel() == 500 for 20000 times torch.double
5.252885223000021
torch.cumsum(a) a.numel() == 700 for 20000 times torch.half
5.310154566999927
torch.cumsum(a) a.numel() == 700 for 20000 times torch.float
5.504942746999859
torch.cumsum(a) a.numel() == 700 for 20000 times torch.double
8.169156298999951
```

After:

```
torch.cumsum(a) a.numel() == 500 for 20000 times torch.half
2.5166763529996388
torch.cumsum(a) a.numel() == 500 for 20000 times torch.float
3.7679748860027757
torch.cumsum(a) a.numel() == 500 for 20000 times torch.double
5.305459807001171
torch.cumsum(a) a.numel() == 700 for 20000 times torch.half
5.373781155998586
torch.cumsum(a) a.numel() == 700 for 20000 times torch.float
5.504503838994424
torch.cumsum(a) a.numel() == 700 for 20000 times torch.double
8.16695938200428
```
---------------------------------------------------------
**Operator : cumprod**
Benchmark with same build settings on same system.
gcc : version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)
CUDA : 10.1
GPU : 1050ti

```python
import timeit

for n, t in [(500, 20000),
             (700, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.cumprod(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.cumprod(a, 0);torch.cumprod(a, 1); torch.cuda.synchronize()',
                              setup=f'import torch; a=torch.randn(({n}, {n}), dtype={dtype}, device="cuda")',
                              number=t))
```

Before:

```
torch.cumprod(a) a.numel() == 500 for 20000 times torch.half
2.7292451850000816
torch.cumprod(a) a.numel() == 500 for 20000 times torch.float
3.7476119149998794
torch.cumprod(a) a.numel() == 500 for 20000 times torch.double
5.308343969000134
torch.cumprod(a) a.numel() == 700 for 20000 times torch.half
5.308035579999796
torch.cumprod(a) a.numel() == 700 for 20000 times torch.float
5.465238159999899
torch.cumprod(a) a.numel() == 700 for 20000 times torch.double
8.13390641900014
```

After:

```
torch.cumprod(a) a.numel() == 500 for 20000 times torch.half
2.5496576320001623
torch.cumprod(a) a.numel() == 500 for 20000 times torch.float
3.7707373259981978
torch.cumprod(a) a.numel() == 500 for 20000 times torch.double
5.320406719998573
torch.cumprod(a) a.numel() == 700 for 20000 times torch.half
5.374952328995278
torch.cumprod(a) a.numel() == 700 for 20000 times torch.float
5.508017359003134
torch.cumprod(a) a.numel() == 700 for 20000 times torch.double
8.198421291999693
```

**Comment by VitalyFedyunin** [comment]
Hello! Can you please separate `logcumsumexp` out of this PR. And synchronize work with #32876 and #26411

**Comment by kshitij12345** [comment]
Sure. Thanks.

**Comment by anjali411** [comment]
@kshitij12345 any updates on the cumprod migration for cuda?

**Comment by kshitij12345** [comment]
@anjali411 I am waiting actually hoping to land #36308 (which brings the reduce cuda kernel). I believe it is mostly ready as the stable gradient computation and relevant tests have been added.

**Comment by anjali411** [comment]
> @anjali411 I was planning to remove the `logcumsumexp` code from this PR, but it has changed a lot since the PR was opened.
> 
> So I am actually hoping to land #36308 (which brings the reduce cuda kernel). I believe it is mostly ready as the stable gradient computation and relevant tests have been added. Post which fixing the conflicts will hopefully be easy.
> 
> Let me know if `cumprod/cumsum` is a priority, will get this ready by Sunday if it is.

Hi @kshitij12345 thanks for the update! that sounds good.

`cumprod` is used in the backward computation of `prod`, which block complex backward for `prod` so it's high priority.

**Comment by ngimel** [comment]
cuda operations are launched asynchronously, you cannot time them with just `timeit.timeit`

**Comment by kshitij12345** [comment]
> cuda operations are launched asynchronously, you cannot time them with just timeit.timeit

@ngimel Wouldn't the `torch.cuda.synchronize()` take care of the synchronising timing.

Like since I run it on same system, both operations will be launched eagerly and asynchronously with the resources available on GPU and then we wait till both are computed. Wouldn't it be okay to use it with `timeit` in this case.

Thanks.

**Comment by ngimel** [comment]
You are calling it at every iteration, forcibly synchronizing, and increasing time, especially for small operations. Usually cuda ops are benchmarked by synchronizing in the beginning and the end of all the benchmarked iterations, to measure actual cuda time rather than cpu overheads surrounding the operation.

**Comment by kshitij12345** [comment]
Benchmark with same build settings on same system.
gcc : version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)
CUDA : 10.1
GPU : 1050ti

```python
import time
import torch
import timeit

for n, t in [(500, 20000),
             (700, 20000)]:
    for dtype in (torch.half, torch.float, torch.double):
        # Input Setup
        a=torch.randn((n, n), dtype=dtype, device="cuda")
        print(f'torch.cumsum(a) a.numel() == {n} for {t} times {dtype}')
        start = time.time()
        # Iterate
        for _ in range(t):
            torch.cumsum(a, 0)
            torch.cumsum(a, 1)
        # Final Synchronize Before Teardown
        torch.cuda.synchronize()
        print("Took:", time.time() - start)
```

# Before:

```
torch.cumsum(a) a.numel() == 500 for 20000 times torch.float16
Took: 2.302349090576172
torch.cumsum(a) a.numel() == 500 for 20000 times torch.float32
Took: 3.5205395221710205
torch.cumsum(a) a.numel() == 500 for 20000 times torch.float64
Took: 5.020417928695679
torch.cumsum(a) a.numel() == 700 for 20000 times torch.float16
Took: 5.088986158370972
torch.cumsum(a) a.numel() == 700 for 20000 times torch.float32
Took: 5.255493640899658
torch.cumsum(a) a.numel() == 700 for 20000 times torch.float64
Took: 8.07828164100647
```

# After:

```
torch.cumsum(a) a.numel() == 500 for 20000 times torch.float16
Took: 2.246642589569092
torch.cumsum(a) a.numel() == 500 for 20000 times torch.float32
Took: 3.5119950771331787
torch.cumsum(a) a.numel() == 500 for 20000 times torch.float64
Took: 5.068270921707153
torch.cumsum(a) a.numel() == 700 for 20000 times torch.float16
Took: 5.115281343460083
torch.cumsum(a) a.numel() == 700 for 20000 times torch.float32
Took: 5.256054878234863
torch.cumsum(a) a.numel() == 700 for 20000 times torch.float64
Took: 7.912769794464111
```

**Comment by kshitij12345** [comment]
@ngimel Does the above method look better?

**Comment by ngimel** [comment]
You'd need a synchronization before iterations, and also it makes sense to split benchmarks for dim=0 and dim=1.

File: aten/src/ATen/native/cuda/ScanKernels.cu [hunk 1]
```diff
@@ -432,4 +432,74 @@ Tensor _logcumsumexp_cuda(const Tensor& self, int64_t dim) {
...
+    return result;
+  }
+  auto wrap_dim = maybe_wrap_dim(dim, self.dim());
+
+  TensorArg output_arg{result, "output", 1};
+  TensorArg input_arg{self, "input", 2};
+  checkAllSameGPU("cumsum", {output_arg, input_arg});
+
+  AT_DISPATCH_ALL_TYPES_AND(
+      at::ScalarType::Half, self.scalar_type(), "cumsum_cuda", [&]() {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ScanKernels.cu:452]: this assumes `self` and `result` have the same datatype, but I don't see it being checked anywhere. When users supply `out` argument it can be of any type.

File: aten/src/ATen/native/cuda/ScanKernels.cu [hunk 2]
```diff
@@ -244,16 +244,16 @@ void cummin_helper_cuda(const Tensor& self, Tensor& values, Tensor& indices, int
  */
 template<typename scalar_t, class BinaryOp>
 __global__ void tensor_kernel_scan_outer_dim(scalar_t *tgt_, scalar_t *src_,
-                                              int num_orows, int num_irows, int row_size,
+                                              unsigned num_orows, unsigned num_irows, unsigned row_size,
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ScanKernels.cu:247]: tensors can, in rare cases, have more rows than 2^32 (or have bigger row_size). It's ok to error out in this case, because using int64_t for indexing incurs perf penalty. But here it will just be a silent error, right?

File: aten/src/ATen/native/cuda/ScanKernels.cu (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/ScanKernels.cu:452]: Makes sense. Will add a check.

File: aten/src/ATen/native/cuda/ScanKernels.cu (hunk 2 above)
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/ScanKernels.cu:247]: Oh. I didn't think of it. The performance was poor with signed int and when I checked with the previous implementation that was a difference.

Yes. Here the value will wrap around.

Is there any operation which errors out on getting row size above 2^32 so that I can check that for reference.

Comment (kshitij12345) [reply aten/src/ATen/native/cuda/ScanKernels.cu:247]: Have added bounds check based on #38970 

Thanks @peterbell10

**Comment by kshitij12345** [comment]
@ngimel PTAL:)

File: aten/src/ATen/native/cuda/ScanKernels.cu [hunk 3]
```diff
@@ -347,21 +347,30 @@ __global__ void tensor_kernel_scan_innermost_dim(T *tgt_, T *src_,
...
+  int64_t num_irows = std::accumulate(sizes.begin() + dim + 1, sizes.end(), 1, std::multiplies<int64_t>());
 
   dim3 threads(std::min(512, int(num_irows)));
-  int maxGridDim = at::cuda::getCurrentDeviceProperties()->maxGridSize[0];
-  dim3 grid(std::min(maxGridDim, num_orows), std::min(maxGridDim, ceil_div(num_irows, int(threads.x))));
+  int64_t maxGridDim = at::cuda::getCurrentDeviceProperties()->maxGridSize[0];
+  dim3 grid(std::min(maxGridDim, num_orows), std::min(maxGridDim, ceil_div(num_irows, int64_t{threads.x})));
+
+  check_fits_in_unsigned(num_irows, "num_irows");
+  check_fits_in_unsigned(num_orows, "num_orows");
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ScanKernels.cu:373]: you should also check row_size?

File: aten/src/ATen/native/cuda/ScanKernels.cu [hunk 4]
```diff
@@ -371,13 +380,16 @@ __host__ void scan_outer_dim(const Tensor& self, Tensor& result,
...
-  int num_rows = self.numel() / row_size;
+  int64_t row_size = self.size(ndim - 1);
+  int64_t num_rows = self.numel() / row_size;
 
   dim3 threads(16, 32);
-  dim3 grid(std::min(at::cuda::getCurrentDeviceProperties()->maxGridSize[0], ceil_div(num_rows, int(threads.y))));
+  int64_t maxGridDim = at::cuda::getCurrentDeviceProperties()->maxGridSize[0];
+  dim3 grid(std::min(maxGridDim, ceil_div(num_rows, int64_t{threads.y})));
+
+  check_fits_in_unsigned(num_rows, "Number of rows (self.numel()/self.size(self.dim()-1))");
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ScanKernels.cu:392]: and check row_size here?

**Review by ngimel — COMMENTED** [review]
Looks great, I have a minor comment/question about row size.

File: aten/src/ATen/native/cuda/ScanKernels.cu (hunk 3 above)
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/ScanKernels.cu:373]: Oops. Had missed that. Thanks for catching that.

File: aten/src/ATen/native/cuda/ScanKernels.cu (hunk 4 above)
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/ScanKernels.cu:393]: Oops. Had missed that. Thanks for catching that.

**Comment by anjali411** [comment]
@kshitij12345 would you like to work on a follow-up PR to add `cumsum` and `cumprod` for complex types? `cumprod` is used in prod_backward so it would be useful to add it for complex

**Comment by kshitij12345** [comment]
@anjali411 Sure. Thank You.
